### PR TITLE
for review

### PR DESCRIPTION
**Pull request by rnetser**

### CODE REVIEW

**Review by dbasunag — APPROVED** [review]

**Comment by rnetser** [comment]
!verified

**Review by myakove — CHANGES_REQUESTED** [review]

File: for_review/file_utils.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    If this runs in the scope of a test the directory path structure will include the test node path
+    If this is run outside the scope of a test the directory path will be for utilities
+
+    Returns:
+        str: test_dir (the directory prefixed for collecting logs)
+    """
+    test_dir = os.environ.get("TEST_DIR_LOG")
+    if not test_dir:
+        # log collection was requested outside the scope of a test
```
[the remark this answers is not in the document]
Comment (rnetser) [reply for_review/file_utils.py:56]: no

File: for_review/infra.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    Args:
+         nodes(list): List of Node objects
+
+    Raises:
+        AssertionError: Assert on node(s) in not ready state
+    """
+    LOGGER.info("Verify all nodes are ready.")
+    not_ready_nodes = [node.name for node in nodes if not node.kubelet_ready]
+    if not_ready_nodes:
+        raise ClusterSanityError(
```
[the remark this answers is not in the document]
Comment (rnetser) [reply for_review/infra.py:26]: why not? cluster is not ready.

File: for_review/must_gather.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def create_must_gather_command(
+    dest_dir,
+    image_url,
+    script_name=None,
+    kubeconfig=None,
+):
+    base_command = (
+        f"oc adm must-gather {f'--kubeconfig {kubeconfig}' if kubeconfig else ''} --image={image_url}"
```
Comment (dbasunag) [inline for_review/must_gather.py:19]: Should the kubeconfig be read from environment variable, if not passed?

File: for_review/file_utils.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+def prepare_test_dir_log_utilities():
+    """
+    Prepares a "utilities" directory under the base log collection directory
+
+    Returns:
+        str: TEST_DIR_LOG (the base directory for log collection)
+    """
+    test_dir_log = os.path.join(
+        os.environ.get("TEST_COLLECT_BASE_DIR"),
```
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/file_utils.py:36]: @myakove seems a good idea, please let us know what should be the default location? How about making $testdir/utilities as a default TEST_COLLECT_BASE_DIR if it's not set?

File: for_review/file_utils.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    If this is run outside the scope of a test the directory path will be for utilities
+
+    Returns:
+        str: test_dir (the directory prefixed for collecting logs)
+    """
+    test_dir = os.environ.get("TEST_DIR_LOG")
+    if not test_dir:
+        # log collection was requested outside the scope of a test
+        test_dir = prepare_test_dir_log_utilities()
+    os.makedirs(test_dir, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/file_utils.py]: Done!

File: for_review/infra.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            err_str=f"Following nodes are not in ready state: {not_ready_nodes}",
+        )
+
+
+def validate_nodes_schedulable(nodes):
```
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/infra.py]: Done!

File: for_review/must_gather.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        f" --dest-dir={dest_dir}"
+    )
+    return f"{base_command} -- {script_name}" if script_name else base_command
+
+
+def run_cnv_must_gather(must_gather_cmd):
```
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/must_gather.py:25]: @myakove, Can you please elaborate?

File: for_review/must_gather.py (hunk 3 above)
Comment (rnetser) [reply for_review/must_gather.py]: No, we should not "hide" it (mainly when working in multi-cluster env)

File: for_review/infra.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (myakove) [reply for_review/infra.py:26]: this is node not ready, not cluster sanity
cluster sanity can call this and raise it's own exception

File: for_review/must_gather.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (myakove) [reply for_review/must_gather.py:25]: why do we need both `create_must_gather_command` and `run_cnv_must_gather` ?

File: for_review/logger.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+def separator(symbol_, val=None):
+    terminal_width = shutil.get_terminal_size(fallback=(120, 40))[0]
+    if not val:
+        return f"{symbol_ * terminal_width}"
+
+    sepa = int((terminal_width - len(val) - 2) // 2)
+    return f"{symbol_ * sepa} {val} {symbol_ * sepa}"
+
+
+def setup_logging(log_level, log_file="/tmp/pytest-tests.log"):
```
Comment (rnetser) [inline for_review/logger.py:38]: we should get name here and use it in line 39
`logger_obj = logging.getLogger(name)`

Comment (adtandale) [reply for_review/logger.py]: Done! Please review.

File: for_review/must_gather.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/must_gather.py:25]: I think create_must_gather_command is for generating/making a command out of which needed command can be made and executed using run_cnv_mus_gather function. Maybe in the future, we may want to make another command using create_must_gather_command which can be used in another new function, that's why we may require both these two functions. @rnetser please correct me if you think differently.

File: for_review/infra.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rnetser) [reply for_review/infra.py:26]: Ack
@adtandale please create new exceptions

File: for_review/file_utils.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
+import logging
+import os
+
+
+LOGGER = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/file_utils.py]: Done!

File: for_review/logger.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+def separator(symbol_, val=None):
+    terminal_width = shutil.get_terminal_size(fallback=(120, 40))[0]
+    if not val:
+        return f"{symbol_ * terminal_width}"
+
+    sepa = int((terminal_width - len(val) - 2) // 2)
+    return f"{symbol_ * sepa} {val} {symbol_ * sepa}"
+
+
+def setup_logging(log_level="info", log_file_name=__name__):
```
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/logger.py]: Done!

File: for_review/file_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/file_utils.py:36]: Done!

File: for_review/infra.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/infra.py:26]: Done!

File: for_review/infra.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+    Raises:
+        AssertionError: Asserts on node(s) not schedulable
+    """
+    LOGGER.info("Verify all nodes are schedulable.")
+    unschedulable_nodes = [
+        node.name for node in nodes if node.instance.spec.unschedulable
+    ]
+    if unschedulable_nodes:
+        raise ClusterSanityError(
```
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/infra.py:46]: Done!

File: for_review/must_gather.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (btjd) [reply for_review/must_gather.py:25]: should this method be renamed from 'run_cnv_must_gather' to just 'run_must_gather'? I take it this is not cnv specific, correct?

File: for_review/file_utils.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+def write_to_file(file_name, content, dir_name="extras"):
+    """
+    This will write to a file that will be available after the test execution,
+
+    Args:
+        file_name (string): name of the file to write
+        content (string): the content of the file to write
+        dir_name (string): (optional) the directory name to create inside the test collect dir
+    """
+    test_dir = collect_logs_prepare_test_dir()
```
[the remark this answers is not in the document]
Comment (rnetser) [reply for_review/file_utils.py:17]: i guess you mean `base_dir`

File: for_review/logger.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+            self.repeated_number = 0
+            return True
+        if repeated_number_exists:
+            self.repeated_number += 1
+        else:
+            self.repeated_number = 1
+        return False
+
+
+def separator(symbol_, val=None):
```
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/logger.py]: Done!

File: for_review/file_utils.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    os.makedirs(extras_dir, exist_ok=True)
+    file_path = os.path.join(extras_dir, file_name)
+    try:
+        with open(file_path, "w") as fd:
+            fd.write(content)
+    except Exception as exp:
+        LOGGER.error(f"Failed to write extras to file: {file_path} {exp}")
+
+
+def prepare_test_dir_log_utilities():
```
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/file_utils.py]: Done!

File: for_review/file_utils.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/file_utils.py]: Done!

File: for_review/file_utils.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    Provides and ensures the creation of a directory to collect logs
+
+    If this runs in the scope of a test the directory path structure will include the test node path
+    If this is run outside the scope of a test the directory path will be for utilities
+
+    Returns:
+        str: test_dir (the directory prefixed for collecting logs)
+    """
+    test_dir = os.environ.get("TEST_DIR_LOG")
```
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/file_utils.py]: Done!

File: for_review/logger.py [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    terminal_width = shutil.get_terminal_size(fallback=(120, 40))[0]
+    if not val:
+        return f"{symbol_ * terminal_width}"
+
+    sepa = int((terminal_width - len(val) - 2) // 2)
+    return f"{symbol_ * sepa} {val} {symbol_ * sepa}"
+
+
+def setup_logging(name, log_level="info", log_file_name="/tmp/pytest-tests.log"):
```
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/logger.py]: Done!

File: for_review/file_utils.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+    Prepares a "utilities" directory under the base log collection directory
+
+    Returns:
+        str: TEST_DIR_LOG (the base directory for log collection)
+    """
+    test_dir_log = os.path.join(
+        os.environ.get("TEST_COLLECT_BASE_DIR"),
+        "utilities",
+    )
+    if os.environ.get("TEST_COLLECT_BASE_DIR") is not None:
```
[the remark this answers is not in the document]
Comment (adtandale) [reply for_review/file_utils.py]: Done!

**Comment by rnetser** [comment]
@adtandale please see the changes introduced in cnv-tests PR #423561 (and https://github.com/RedHatQE/openshift-python-wrapper/pull/733 )

Please update the code with the generic things in 423561 (and remove reference to tests).

@myakove ^^

**Comment by adtandale** [comment]
Verified code written in for_review/file_utils.py

**Comment by adtandale** [comment]
verified code in infra.py and exceptions.py

**Review by rnetser — COMMENTED** [review]
Please update the PR

File: for_review/infra.py [hunk 18]
```diff
@@ -0,0 +1,87 @@
+import logging
+import subprocess
+
+import kubernetes as kubernetes
+from kubernetes.dynamic import DynamicClient
+
+from ocp_utilities.exceptions import NodeNotReadyError, NodeUnschedulableError
+
+
+LOGGER = logging.LOGGER = logging.getLogger(name=__name__)
```
Comment (rnetser) [inline for_review/infra.py:10]: should use setup_logging

File: for_review/logger.py [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+            self.repeated_number = 0
+            return True
+        if repeated_number_exists:
+            self.repeated_number += 1
+        else:
+            self.repeated_number = 1
+        return False
+
+
+class TestLogFormatter(ColoredFormatter):
```
Comment (rnetser) [inline for_review/logger.py:30]: please rename anything with "test" to something else (in all places)

File: for_review/exceptions.py [hunk 20]
```diff
@@ -0,0 +1,22 @@
+class ClusterSanityError(Exception):
+    def __init__(self, err_str):
```
Comment (dbasunag) [inline for_review/exceptions.py:2]: Are these needed? We can simply call ClusterSanityError(), with error string.

File: for_review/infra.py [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+
+def assert_nodes_schedulable(nodes):
+    """
+    Validates all nodes are in schedulable state
+
+    Args:
+         nodes(list): List of Node objects
+
+    Raises:
+        AssertionError: Asserts on node(s) not schedulable
```
Comment (dbasunag) [inline for_review/infra.py:39]: Ditto.

File: for_review/infra.py [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+    """
+    LOGGER.info("Verify all nodes are ready.")
+    not_ready_nodes = [node.name for node in nodes if not node.kubelet_ready]
+    if not_ready_nodes:
+        raise NodeNotReadyError(
+            err_str=f"Following nodes are not in ready state: {not_ready_nodes}",
+        )
+
+
+def assert_nodes_schedulable(nodes):
```
Comment (dbasunag) [inline for_review/infra.py:31]: Ditto.

**Review by dbasunag — CHANGES_REQUESTED** [review]

File: for_review/infra.py [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+import kubernetes as kubernetes
+from kubernetes.dynamic import DynamicClient
+
+from ocp_utilities.exceptions import ClusterSanityError
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+def validate_nodes_ready(nodes):
```
[the remark this answers is not in the document]
Comment (rnetser) [reply for_review/infra.py:14]: keeping the name as we raise now

File: for_review/infra.py [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+import kubernetes as kubernetes
+from kubernetes.dynamic import DynamicClient
+
+from ocp_utilities.exceptions import NodeNotReadyError, NodeUnschedulableError
+
+
+LOGGER = logging.LOGGER = logging.getLogger(name=__name__)
+
+
+def assert_nodes_ready(nodes):
```
[the remark this answers is not in the document]
Comment (rnetser) [reply for_review/infra.py:11]: should use "assert" in name if we raise

**Comment by rnetser** [comment]
!verified

**Review by myakove — APPROVED** [review]

**Review by myakove — CHANGES_REQUESTED** [review]
please fix

**Review by myakove — APPROVED** [review]

**Review by myakove — COMMENTED** [review]
test

**Review by myakove — COMMENTED** [review]
test

**Review by myakove — APPROVED** [review]

**Comment by rnetser** [comment]
!verified

**Review by dbasunag — APPROVED** [review]

**Review by myakove — APPROVED** [review]

**Pull request merged**